pubtools/pulplib/_impl/model/repository/file.py

```diff
@@ -20,3 +30,77 @@ class FileRepository(Repository):
     )
 
     mutable_urls = attr.ib(default=attr.Factory(lambda: ["PULP_MANIFEST"]), type=list)
+
+    def upload_file(self, file_obj, relative_url=None):
+        """Upload a file to this repository.
+
+        Args:
+            file_obj (str, file object)
+                If it's a string, then it's the path of file to upload.
+                Else, it ought to be a file-like object, see:
+                https://docs.python.org/3/glossary.html#term-file-object
+
+            relative_url (str)
+                Path that should be used in remote repository, can either
+                be a path to a directory or a path to a file, e.g:
+                - if relative_url is 'foo/bar' and file_obj has name 'f.txt',
+                  the result remote path wll be 'foo/bar/f.txt'.
+                - if relative_url is 'foo/bar/f.txt', no matter what the
+                  name of file_obj is, the remote path is 'foo/bar/f.txt'.
+
+                If omitted, the local name of the file will be used. Or,
+                if file_obj is a file object without a `name` attribute,
+                passing `relative_url` is mandatory.
+
+        Returns:
+            Future[:class:`~pubtools.pulplib.Task`]
+                A future which is resolved when import succeeds.
+
+                The future contains the task to import uploaded content
+                to repository.
+
+        Raises:
+            DetachedException
+                If this instance is not attached to a Pulp client.
+
+        .. versionadded:: 1.2.0
+        """
+        if not self._client:
+            raise DetachedException()
+
+        relative_url = self._get_relative_url(file_obj, relative_url)
+
+        # request upload id and wait for it
+        upload_id = self._client._request_upload().result()["upload_id"]
+
+        upload_complete_f = self._client._do_upload_file(upload_id, file_obj)
+
+        import_complete_f = f_flat_map(
+            upload_complete_f,
+            lambda upload: self._client._do_import(
+                self.id,
+                upload_id,
+                "iso",
+                {"name": relative_url, "digest": upload[0], "size": upload[1]},
+            ),
+        )
+
+        f_map(
+            import_complete_f, lambda _: self._client._delete_upload_request(upload_id)
+        )
+
+        return import_complete_f
+
+    def _get_relative_url(self, file_obj, relative_url):
+        is_path = isinstance(file_obj, six.string_types)
```

Comment:
As mentioned elsewhere, I still don't think this is the right way to decide how to handle file_obj.

Also, you should probably encapsulate that in a helper function so you don't have to write the same code (fix the same bugs) twice.